### ISSUE TITLE
Conform to XDG BASE DIRECTORY specification

### ISSUE THREAD

**Issue by y0shir** [TheXTech/TheXTech#395]
Freedesktop XDG Base Directory specification states that applications should use path saved in $XDG_DATA_HOME for user data and $XDG_CONFIG_HOME for configuration and in case those are not set  $HOME/.local/share and $HOME/.config .

**Comment by Wohlstand** [comment]
Good note, I'll tweak my DEB/RPM package builds and then forward user directories to these locations.

**Comment by ChristianSilvermoon** [comment]
Reference (based on current implementation):
- `$XDG_CONFIG_HOME` defaulting to `$HOME/.config` if not set
- - `gameinfo.ini` (If not tied to an asset pack)
- - `sounds.ini` (If not tied to an asset pack)
- `$XDG_STATE_HOME` defaulting to `$HOME/.local/state` if not set
- - Logs
- `$XDG_DATA_HOME` defaulting to `$HOME/.local/share` if not set
- - Episodes
- - Levels
- - Asset Packs
- - Saves

[The Latest Spec can be found here](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html)

This should be the default for non-portable Linux Builds and other platforms using XDG Standards

**Comment by Wohlstand** [comment]
> gameinfo.ini

It's the part of assets pack, not user directory

> sounds.ini / music.ini

Also, they are part of the config pack.

All other files at "settings" directory are part of user directory among with logs, screenshots, GIF records, etc.

**Comment by Wohlstand** [comment]
For me, all these variables are undefined:
![изображение](https://user-images.githubusercontent.com/6751442/186988918-f26d5cb0-8799-4dd5-a834-80dd87a96c1a.png)

**Comment by Wohlstand** [comment]
Anyway, making anything with this, I better after the 1.3.6 release, probably at 1.3.6.1 as a small change with something also like bugfixes.

**Comment by ChristianSilvermoon** [comment]
> For me, all these variables are undefined

That's normal :P

Programs are supposed to assume default values if undefined (IE if `$XDG_DATA_HOME` is undefined, it should be treated as `$HOME/.local/share`)

But with this logic, if you were to do 
```bash
# Export for JUST thextech
XDG_DATA_HOME=/media/$USER/somedrive/data XDG_CONFIG_HOME=/media/$USER/somedrive/config XDG_STATE_HOME=/tmp thextech

# OR, export for all programs and launch
export XDG_DATA_HOME=/media/$USER/somedrive/data
export XDG_CONFIG_HOME=/media/$USER/somedrive/config
export XDG_STATE_HOME=/tmp
thextech
```

the game *should* do the following:
* For Episodes, Levels, Asset Packs, Saves, etc use `/media/$USER/somedrive/data/thextech`
* For Settings/Config use `/media/$USER/somedrive/config/thextech`
* For logging, use `/tmp/thextech`

Specification suggests logs belong in `$XDG_STATE_HOME` due to being less portable and/or important to the user, and for having to do with application state rather than user data.
>   The $XDG_STATE_HOME contains state data that should persist between (application) restarts, but that is not important or portable enough to the user that it should be stored in $XDG_DATA_HOME. It may contain:
>
>    * actions history (logs, history, recently used files, …)
>
>    * current state of the application that can be reused on a restart (view, layout, open files, undo history, …)

**Comment by ChristianSilvermoon** [comment]
> It's the part of assets pack, not user directory

Asset Packs can count as user data if user managed though, as in the case of Terraria where resource packs that change the game's look and sound are stored under `$XDG_DATA_HOME/Terraria/ResourcePacks`

I'd assume `$XDG_CONFIG_HOME` is most appropriate only for button mappings, window resolution settings and various 'options'-related things; everything else default to `$XDG_DATA_HOME`

Think of it like `/etc` (Config) vs `/usr/share` (Data)

**Comment by ChristianSilvermoon** [comment]
Additional thought to contribute.

I've done some thinking and I believe that just using `$XDG_DATA_HOME/thextech` or `$XDG_DATA_HOME/thextech-game` would be the simplest/most appropriate approach

**Comment by Wohlstand** [comment]
Just now I made a thing:
https://github.com/Wohlstand/TheXTech/commit/17f5ebd29d4828d453bfc2890e98ce98556dc178

There is some work is still needed. So, what I made (for debug assets):
- `$HOME/.config/thextech/` - for settings and gamesaves
- `$HOME/.local/state/thextech/` - for logs
- `$HOME/.local/share/thextech/debug-assets` - for debug assets
- `$HOME/.local/share/thextech/userdata` - for userdata

**Comment by Wohlstand** [comment]
Also! If legacy debug direcories do exist, they should work as usually, we don't want to break existing environments.

**Comment by Wohlstand** [comment]
So, if speak about default behaviour, the work done. When speak about DEB/RPM releases of games, I will go to split the thing into:
- game itself (no assets)
- assets-only package that depends on a game, and it will supply the BASH script that will set all paths, including checking of XDG paths.

**Comment by ChristianSilvermoon** [comment]
> Just now I made a thing: [17f5ebd](https://github.com/Wohlstand/TheXTech/commit/17f5ebd29d4828d453bfc2890e98ce98556dc178)
> 
> There is some work is still needed. So, what I made (for debug assets):
> 
>     * `$HOME/.config/thextech/` - for settings and gamesaves
> 
>     * `$HOME/.local/state/thextech/` - for logs
> 
>     * `$HOME/.local/share/thextech/debug-assets` - for debug assets
> 
>     * `$HOME/.local/share/thextech/userdata` - for userdata
> 
> 
> `XDG_*` variables will also work.

Since the user may want to override the intro/outro levels with something they made... could it be possible to have introset in userdata append those in debug-assets... and if an intro.lvl or outro.lvl is present in userdata have them override those in debug-assets?

I'm also seemingly not getting sound out and exiting the game leaves the process hanging, forcing me to kill -9 :thinking:

All-in-all I'm definitely loving this it feels so much cleaner!
But I would put gamesaves under $XDG_DATA_HOME since it's your save data rather than your config data

Perhaps at `$XDG_DATA_HOME/thextech/saves` ?

**Comment by Wohlstand** [comment]
Damn!
I accidentally broke the portable mode... Going to fix as soon as possible...

**Comment by Wohlstand** [comment]
> I'm also seemingly not getting sound out and exiting the game leaves the process hanging, forcing me to kill -9 🤔

There was some small crash bug that I fixed somewhere ago... The same happen on macOS. At me it just crashed and closed immediately.

**Comment by Wohlstand** [comment]
>  But I would put gamesaves under $XDG_DATA_HOME since it's your save data rather than your config data

Added `$HOME/.local/share/thextech/gamesaves` - for saves of game. Please verify.

**Comment by Wohlstand** [comment]
I think, the last thing left is an ability to supply the custom "sub-directory name" from command line to let bash scripts know what game they should run, and where engine shall find assets. By the way, I'll add additional assets search path as a fallback - into the "/usr/share/games/<gamename>/" directory. So, with this, will no need to embed that into different copies of executables.

**Comment by Wohlstand** [comment]
Okay, just now I added the `--game-dirname` argument which will allow replace default "thextech" with another name, like, `adventures-of-demo`, and you'll have the next result:
![изображение](https://github.com/Wohlstand/TheXTech/assets/6751442/5d0fc684-d0da-458d-a2d6-bd4cacb87957)

Again, if there is no "debug-assets" sub-directory, the /usr/share/games/ will be used as an assets directory.

**Comment by ds-sloth** [comment]
I'm concerned that this may be too complex. It might be better to exclusively use the debug-assets and userdata folders as @ChristianSilvermoon suggested last week, without separate gamesave and logs folders.

**Comment by ChristianSilvermoon** [comment]
@Wohlstand Tested and the new game saves location is working, but still no audio beyond the initial sound effect that plays when the game is loading and the process still hangs on exit

**Comment by Wohlstand** [comment]
> @Wohlstand Tested and the new game saves location is working, but still no audio beyond the initial sound effect that plays when the game is loading and the process still hangs on exit

I guess, I know why that happen: wrong permissions, I got that too. Please apply `chmod 644` to all files.

**Comment by ChristianSilvermoon** [comment]
> > @Wohlstand Tested and the new game saves location is working, but still no audio beyond the initial sound effect that plays when the game is loading and the process still hangs on exit
> 
> I guess, I know why that happen: wrong permissions, I got that too. Please apply `chmod 644` to all files.

Even though the program is running as my user and the files are owned by the same user? This is with a binary I built myself :thinking:

**Comment by Wohlstand** [comment]
> Even though the program is running as my user and the files are owned by the same user? This is with a binary I built myself 🤔

I meant resources placed at `/usr/share/`, there was a mistake in default permissions that caused a mess.

**Comment by Wohlstand** [comment]
> I'm also seemingly not getting sound out and exiting the game leaves the process hanging, forcing me to kill -9 🤔

Try to run with gdb and tell me, what it says? What backstack?

**Comment by ds-sloth** [comment]
I'd like to remove this from milestone 1.3.6.1 so we can continue discussing these issues and how they interact with #291 more slowly.

@Wohlstand, can you please make sure that current main builds use the legacy user data directories so that users only need to migrate once?

**Comment by Wohlstand** [comment]
Just now I made new paths system being disabled by default:
https://github.com/Wohlstand/TheXTech/commit/8f48fb30fe1acba1214c25603a304787cb1ae42c

That should give us more time to experiment with this, choose the best variant of this, etc. In the 1.3.6.1 the current paths system will be used.

**Comment by ChristianSilvermoon** [comment]
Update:

For simplicity's sake, I think we may want to consider just relying on `$XDG_DATA_HOME/thextech` as the main directory, much like Terraria uses `$XDG_DATA_HOME/Terraria`

**Comment by ds-sloth** [comment]
I personally agree, please share these comments at #732, and feel free to review there if you like.

**Comment by ds-sloth** [comment]
This has been implemented since #732
